### PR TITLE
Fix ocamltest / Windows / Unicode issues

### PR DESCRIPTION
**Pull request by nojb**
This is a first PR to fix some of the CI errors under Windows following the merge of #1200 and #681. ~~It is not yet ready for merging.~~

- The first commit makes `ocamltest` a Unicode Windows application.
- The second commit fixes a bug in #1200. (I think due to a bad rebase; ~~there is an outstanding `FIXME` that I am working on right now.~~)

Also, some "new" tests are still failing in the `compare-native-programs` step.  I am investigating.

/cc @shindere @dra27

### CODE REVIEW

**Comment by nojb** [comment]
This PR seems to fix all tests under Windows, except for the `compare-native-programs` tests of `ocamltest`.  After some detective work with the help of @alainfrisch it seems that the reason is that the PE/COFF  executables embed timestamps (in seconds).  This makes the test fail **sometimes**.

One possibility would be to try to skip the portion of the file that contains the timestamp when comparing them, but it may not be so easy because the timestamp appears in more than one place in the executables.

**Comment by gasche** [comment]
ocamltest already hardcodes the fact that the 512 first bytes of the native programs are ignored under windows: https://github.com/ocaml/ocaml/blob/635b5ed/ocamltest/builtin_actions.ml#L707-L710 . A very conservative approach would be to simply disable this test/action under the same OS conditions.

**Comment by nojb** [comment]
Another possibility is to count the number of different bytes (for example, with `cmp -l FILE1 FILE2 | wc -l`) and succeed as long as this number is not too large.

**Comment by gasche** [comment]
(Apparently there exists a reasonably portable tool, [Ducible](https://github.com/jasonwhite/ducible), that knows how to strip non-reproducible information from PE executables. I'm pointing it out, but not suggesting to use it from ocamltest, because it adds a lot to the dependency chain (it's 10K lines of C++, and depends on Python 3 for configuration/building).

**Comment by nojb** [comment]
Is there a way to trigger AppVeyor on this pull request ?

**Comment by gasche** [comment]
I will try to close and reopen.

A few remarks by having just a very quick look at the patch:

- some changes are not ocamltest-related at all (you say it's a #1200 merge issue) and I find it a bit confusing to have both kinds of changes in the same PR. Could you send the non-ocamltest-related changes in a separate PR? (In particular, those should not need Sébastien's attention/review).

- Apologies for the very naive comment (I know nothing of your Windows-Unicode work), but I find it a bit strange that `ocamltest/run_stubs.c` needs to be changed to use `..._utf16` function, as I assume that this file (unlike `run_win32.c`) is not Windows-specific. Does that mean that arbitrary OCaml C code will have to use `..._utf16` functions for portability now? Is there a guide on porting applications somewhere? Or is that ocamltest-specific and why?

- the implementation of disabling native tests on Windows is morally wrong, as you just overloaded the conditional that shows "tests are skipped under flambda" (so after your patch this line will show under non-flambda Windows builds); you should use a new conditional branch with its own meaningful message.

**Comment by gasche** [comment]
(Apparently close-reopen didn't work, I'm not sure why AppVeyor does not work here. Did we recently break the appveyor metadata files? (But they were last touched by #1200 and AppVeyor kept working after we merged that). Did AppVeyor automatically shut off after too many failed build? cc: @avsm)

**Comment by nojb** [comment]
Thanks @gasche for the comments.

- OK, will open a separate PR for the #1200 fix.

- Indeed, C code that is shared between Unix and Windows needs to use the `_utf16` functions.  In the Unix case these are just defined to be the usual functions, see [here](https://github.com/ocaml/ocaml/blob/trunk/byterun/caml/misc.h#L230-L232).  Do you think it is worth introducing a new name for this three functions to be used in C code shared between Unix and Windows ?

- Indeed, I forgot to fix the message.

**Comment by gasche** [comment]
Personally I find it fairly confusing to have to use `_utf16` functions when my own system doesn't use utf16 anywhere. I would rather expect `caml_stat_strdup` to do the right thing on systems I don't know about, rather than me learning about the function names they need. But maybe that is too naive and cannot work. Has this been discussed before? Otherwise I would recommend starting a discussion with the development team on an appropriate place -- maybe in the comments for #1200? Or in a [Discourse](https://discuss.ocaml.org/) thread?

**Comment by gasche** [comment]
(I think it's better to have two separate tests with two precise error messages, one for flambda and one for windows, rather than a message that enumerates various possible skipping causes.)

**Comment by nojb** [comment]
OK, amended the `ocamltest` message as suggested and removed unrelated commits, except for one tiny fix to `tests/letrec` test to avoid a warning when running the testsuite (let me know if I should remove that one as well).

**Comment by shindere** [comment]
Hi, sorry for the late response, will address all issues ASAP.
@nojb I don't know what makes AppVeyor restart its builds. In my
experience it does so each time a branch on which a PR has been opened
is updated. If you don't already have an AppVeyor account you may try to
create one.

Or perhaps this behaviour is because I'm part of the developers for the
ocaml project, I have no idea.

**Comment by alainfrisch** [comment]
> Personally I find it fairly confusing to have to use _utf16 functions when my own system doesn't use _utf16 anywhere.

I tend to agree.  Perhaps just changing the suffix from _utf16 to _sys (or something else) would make it clearer that this is about converting between OCaml strings and the ones expected/returned by the OS (the identity and same types on both sides under \*nix, the proper conversion and char* vs wchar* under Windows).

**Comment by nojb** [comment]
OK, I will open a new PR with this change and we can continue the discussion there.

**Comment by shindere** [comment]
Hi, NIcolas,

Many thanks for all your work on Unicode support for Windows in general
and ocamltest in particular.

Am I correct that the intention is that #1363 gets merged first and then
the present PR gets rebased on it so that run_stubs does not need to
mention utf16 any longer?

Also, with your PR I noticed that run_main.c leaked in although i didn't
intend to publish it. There is also a comment I'd like to get rido of in
the function that handles string arrays. Shall I do a PR for these two
things and let you rebase yours on mine? Or shall I push to your branch
directly? Or shall this PR get merged first and then I do a PR to update
the other things?

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/19 16:55 +0000):
> This PR seems to fix all tests under Windows, except for the
> `compare-native-programs` tests of `ocamltest`.  After some detective
> work with the help of @alainfrisch it seems that the reason is that
> the PE/COFF  executables embed timestamps (in seconds).  This makes
> the test fail **sometimes**.

Thanks! Actually, I was aware of the timestamp issue, it is something we
discussed with @xavierleroy. So I started by going through the COFF
documentation. My hope was to be able to figure out exactly where this
timestamp is to be able to skip the precise number of bytes. Since I
failed to do that in a proper way, I hard-coded a skip of 512 bytes
which seemed to be enough when I tested last year but now seems not
appropriate any longer.

> One possibility would be to try to skip the portion of the file that
> contains the timestamp when comparing them, but it may not be so easy
> because the timestamp appears in more than one place in the
> executables.

Ah I didn't know it appears at several places. But I did attempt to do
what you suggest, as noticed by @gasche.

**Comment by shindere** [comment]
FWIW, in the days just before #681 was merged, I had to disable the
comparison of native binaries in the case where flambda is enabled.
Indeed, on one test file (basic-more/morematch.ml IIRC) the binaries
produced by ocamlopt.byte and ocamlopt.opt (under Linux) with flambda
enabled were different.

I don't know if this can be fixed, but at the very least this is
something I'd like to understand better. Any idea @chambart?

The file mentionned above did also produce different binaries when
compiled with ocamlopt.byte and ocamlopt.opt when the level of inlining
was >=5.

**Comment by nojb** [comment]
Hi Sébastien,

Either way is fine for me. I can easily rebase the PRs if needed.  I propose thus

1) you make a PR with your two little fixes
2) I rebase this PR on that
3) we try to merge this PR without waiting for #1363, since this PR is necessary to fix the Windows tests and #1363 is "cosmetic".

**Comment by nojb** [comment]
@shindere Out of curiosity did you try investigating the binary diff of `morematch` under both compilers ?

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/20 04:12 -0700):
> Another possibility is to count the number of different bytes (for
> example, with `cmp -l FILE1 FILE2 | wc -l`) and succeed as long as
> this number is not too large.

That's an interesting idea, thanks.

Well to be honest I really like the idea of comparing the native
binaries and I would love to be able to do that in a proper way. Partly
for aesthetic reasons, but also because I think the more properly it is
done, the more robust and sustainable it will be. Skipping the beginning
of the files or making sure the number of differences is not too big
sounds less satisfactory, although I fully realise it may be the best
one can reasonably expect.

Disabling the comparison for Windows native binaries completely is also
an option, of course. I don't know what's our best choice here. Any
opinion welcome.

**Comment by shindere** [comment]
Many thanks for your careful review, @gasche !

**Comment by nojb** [comment]
Well, another possibility is to teach `ocamltest` a teeny bit of COFF so that it can step over the timestamps. It sounds like a fun exercise.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/21 07:04 -0700):
> @shindere Out of curiosity did you try investigating the binary diff
> of `morematch` under both compilers ?

No, due to a lack of time. But I hope I'll be able to investigate now
that ocamltest has finally been merged.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/21 14:14 +0000):
> Well, another possibility is to teach `ocamltest` a teeny bit of COFF
> so that it can step over the timestamps. It sounds like a fun
> exercise.

Yeah I started to do it. Look for the comparison_start_address function
in ocamltest/builtin_actions.ml. It's just that it didn't work, most
likely because I misread the spec.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/21 06:57 -0700):
> Hi Sébastien,
> 
> Either way is fine for me. I can easily rebase the PRs if needed.  I propose thus
> 
> 1) you make a PR with your two little fixes
> 2) I rebase this PR on that
> 3) we try to merge this PR without waiting for #1363, since this PR is
> necessary to fix the Windows tests and #1363 is "cosmetic".

Thanks! See #1364.

**Comment by nojb** [comment]
OK, I think I understand why AppVeyor is not working on any of my PRs: I based them off `4.06` instead of `trunk`.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/21 14:46 +0000):
> OK, I think I understand why AppVeyor is not working on any of my PRs:
> I based them off `4.06` instead of `master`.

Watch out! It should really be trunk, not master!

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/21 14:46 +0000):
> OK, I think I understand why AppVeyor is not working on any of my PRs:
> I based them off `4.06` instead of `master`.

Btw I tend to *always* rebase on latest trunk.

**Comment by nojb** [comment]
Rebased on `trunk`.

**Comment by gasche** [comment]
> Disabling the comparison for Windows native binaries completely is also
> an option, of course. I don't know what's our best choice here. Any
> opinion welcome.

My opinion: you could take the shortest route to having working tests without false negatives (that is, disable comparison of window native binaries for now), and then later iterate to refine to test more. (If the second part is more difficult and only comes after the 4.06 release, I personally think it is fine.)

**Comment by shindere** [comment]
Gabriel Scherer (2017/09/21 08:43 -0700):
> > Disabling the comparison for Windows native binaries completely is also
> > an option, of course. I don't know what's our best choice here. Any
> > opinion welcome.
> 
> My opinion: you could take the shortest route to having working tests
> without false negatives (that is, disable comparison of window native
> binaries for now), and then later iterate to refine to test more. (If
> the second part is more difficult and only comes after the 4.06
> release, I personally think it is fine.)

Sure, would be okay with this.

**Comment by xavierleroy** [comment]
Speaking of ocamltest on Windows, the CI at Inria is failing on the MSVC and Mingw "native Windows" ports with errors in several (many?) of the tests that have been converted to use ocamltest:
```
List of failed tests:
    tests/basic-modules/'main.ml' with 1 (native) 
    tests/basic-io-2/'io.ml' with 1 (native) 
    tests/basic-multdef/'usemultdef.ml' with 2 (bytecode) 
    tests/basic-io/'wc.ml' with 2 (bytecode) 
    tests/basic-io/'wc.ml' with 1 (native) 
    tests/basic-modules/'main.ml' with 2 (bytecode) 
    tests/array-functions/'test.ml' with 2 (bytecode) 
    tests/basic-io-2/'io.ml' with 2 (bytecode) 
    tests/array-functions/'test.ml' with 1 (native) 
    tests/basic-multdef/'usemultdef.ml' with 1 (native) 
    tests/basic-private/'tlength.ml' with 1 (native) 
    tests/basic-private/'tlength.ml' with 2 (bytecode) 
```
I compiled a couple of those tests by hand and their output was correct modulo CR-LF/LF equivalence.

Does anyone else observes this problem?  How do we get ocamltest to explain the failure in more details?

**Comment by nojb** [comment]
I am not sure about the Inria CI, but a subset of those failures are also visible in AppVeyor CI (https://ci.appveyor.com/project/avsm/ocaml/build/1.0.4266#L10321) and I believe they are fixed by #1362.

**Comment by shindere** [comment]
To see why a test fails, see the _ocamltest subdirectory. That's where
ocamltest write its log and also the results of compiling and running
tests.

**Comment by shindere** [comment]
It seems that, at least for some tests, it is CreateProcess which fails.
I can see only one letter for the error message, "T". Can't investigate
more right now but perhaps this will give someody a hint?

**Comment by nojb** [comment]
Only one letter being printed - seems like a `WCHAR *` is being passed to a function expecting a `char *`.  Will look into it.

**Comment by shindere** [comment]
It's also ocamltest which does a too poor job at checking exactly what
happens when CreateProcess fails. Will improve this ASAP, hopefully
tonight, and hopefully I can give more useful feedback.

**Comment by shindere** [comment]
@nojb well actually the support for error messages has been implemented
in ocamltest but perhaps it is not done well enough regarding unicdoe
support.

Could you please have a look to the report_error function in run_win32.c
and see whether it looks good to you?

Thanks!

**Comment by nojb** [comment]
I cannot reproduce the `native` test failures in my machine. In any case, I added a commit that may fix the error message. The bytecode failures should be fixed by #1362.

If someone could post the `_ocamltest` logs (or, even better, a `.zip` with the whole directory) produced by doing `make new` in `testsuite/` in Inria's CI that may help debugging.

**Comment by nojb** [comment]
One guess about the `native` failures in Inria CI: an issue with `flexdll`.

Are you using the binary 0.36 release ? or the bootstrapped version ? I remember running into some issues when I tried to use the bootstrapped version.

**Comment by nojb** [comment]
@shindere I think `report_error` is essentially correct except for the use of `%S` to print wide strings which is pretty broken on Windows (we are looking into this with @dra27).  Instead, in the last commit I explicitly convert the wide string into UTF-8 which should make it work.

File: ocamltest/run_stubs.c [hunk 1]
```diff
@@ -78,17 +79,21 @@ CAMLprim value caml_run_command(value caml_settings)
...
-  settings.stderr_filename = String_val(Field(caml_settings, 4));
+  settings.stdin_filename = caml_stat_strdup_to_utf16(String_val(Field(caml_settings, 2)));
+  settings.stdout_filename = caml_stat_strdup_to_utf16(String_val(Field(caml_settings, 4)));
+  settings.stderr_filename = caml_stat_strdup_to_utf16(String_val(Field(caml_settings, 4)));
   settings.append = Bool_val(Field(caml_settings, 5));
   settings.timeout = Int_val(Field(caml_settings, 6));
   settings.logger = logToChannel;
   settings.loggerData = Channel(Field(caml_settings, 7));
   res = run_command(&settings);
+  caml_stat_free((charnat *)settings.program);
```
Comment (xavierleroy) [inline ocamltest/run_stubs.c]: Why this cast (and several other below) to `charnat *` ? `caml_stat_free` takes `void *` arguments.

File: ocamltest/run_win32.c [hunk 2]
```diff
@@ -26,33 +26,38 @@
...
+  error_message_c = caml_stat_strdup_of_utf16(error_message);
   if ( is_defined(argument) )
     error_with_location(file, line,
-      settings, "%s %s: %s", message, argument, error_message);
+      settings, "%s %s: %s", message, argument, error_message_c);
   else
     error_with_location(file, line,
-      settings, "%s: %s", message, error_message);
+      settings, "%s: %s", message, error_message_c);
+  caml_stat_free(error_message);
```
Comment (xavierleroy) [inline ocamltest/run_win32.c]: AAAARGH!  You're freeing the wrong error_message!!

**Review by xavierleroy — CHANGES_REQUESTED** [review]
One serious error spotted.  The rest looks reasonable to me but could use a second reading.

**Comment by xavierleroy** [comment]
> One guess about the native failures in Inria CI: an issue with flexdll.

That was the case until recently, but now the whole system builds and runs the test suite, it's just the ocamltest-administered tests that report errors.  

I'm confident that the changes from the present PR will address this issue;. Let's just have them reviewed and finished as quickly as possible so that we can merge and try again.

File: ocamltest/run_win32.c (hunk 2 above)
Comment (nojb) [reply ocamltest/run_win32.c]: Thanks, how embarrassing! (there should be a warning somewhere about writing C with less than 5 hours of sleep...)

File: ocamltest/run_stubs.c (hunk 1 above)
Comment (nojb) [reply ocamltest/run_stubs.c]: Yes, this is confusing: the right fix is not to define the struct fields (see https://github.com/ocaml/ocaml/blob/trunk/ocamltest/run.h#L32-L34) as being `const char *` in the original code. Will fix.

**Comment by nojb** [comment]
Fixed and rebased.

**Review by damiendoligez — APPROVED** [review]
I've reviewed the diff and I think it's good.
@xavierleroy you'll need to change your review status...

**Review by xavierleroy — APPROVED** [review]

**Comment by shindere** [comment]
Could not go very far with the tests but I think it would be great to
have this PR merged and perhaps also #1363. Indeed, the interaction between
ocamltest (#681) and Unicode support for Windows (#1200) could not be tested
very well and I think several of the failures that are actually encountered
on Inria's CI come from that. For instance, even the names of the files that
are created to store compiler outputs look weird.

So, my suggestion would be to merge these PRs if they are okay and I expect
the number of failures on Inria's CI to decrease. Then we can see what
errors are left and not due to this interaction.

**Comment by gasche** [comment]
Merged.

**Comment by gasche** [comment]
Note that #1363 conflicts with the present PR, if I understand correctly (the present one uses the pre-1363 names). I'll ask @nojb to rebase taking this code into account.

**Comment by nojb** [comment]
Thanks! Will rebase right away!

**Comment by gasche** [comment]
Cherry-picked in 4.06 [ 9aebb961257fb131aea66196794b4f2c75571578 ]. (Cherry-pick squashed so history is lost, but (1) it was not that clean anyway and (2) this is just in the release branch, trunk has full history.)

**Pull request merged**